pyscriptjs/src/components/pyrepl.ts

```diff
@@ -5,161 +5,201 @@ import { Compartment, StateCommand } from '@codemirror/state';
+            }
+            if (!this.hasAttribute('root')) {
+                this.setAttribute('root', this.id);
+            }
+
+            const pySrc = htmlDecode(this.innerHTML).trim();
             this.innerHTML = '';
-            const languageConf = new Compartment();
+            this.editor = this.makeEditor(pySrc);
+            const boxDiv = this.makeBoxDiv();
+            this.appendChild(boxDiv);
+            this.editor.focus();
+            logger.debug(`element ${this.id} successfully connected`);
+        }
 
+        /** Create and configure the codemirror editor
+         */
+        makeEditor(pySrc: string): EditorView {
+            const languageConf = new Compartment();
             const extensions = [
                 indentUnit.of("    "),
                 basicSetup,
                 languageConf.of(python()),
                 keymap.of([
                     ...defaultKeymap,
-                    { key: 'Ctrl-Enter', run: createCmdHandler(this) },
-                    { key: 'Shift-Enter', run: createCmdHandler(this) },
+                    { key: 'Ctrl-Enter',  run: this.execute.bind(this) },
+                    { key: 'Shift-Enter', run: this.execute.bind(this) },
                 ]),
             ];
 
-            if (getEditorTheme(this) === 'dark') {
+            if (getAttribute(this, 'theme') === 'dark') {
                 extensions.push(oneDarkTheme);
             }
 
-            this.editor = new EditorView({
-                doc: this.code.trim(),
+            return new EditorView({
+                doc: pySrc,
                 extensions,
-                parent: this.editorNode,
             });
+        }
 
-            const mainDiv = document.createElement('div');
-            addClasses(mainDiv, ['py-repl-box']);
+        // ******** main entry point for py-repl DOM building **********
+        //
+        // The following functions are written in a top-down, depth-first
+        // order (so that the order of code roughly matches the order of
+        // execution)
+        makeBoxDiv(): HTMLElement {
+            const boxDiv = document.createElement('div');
+            boxDiv.className = 'py-repl-box';
 
-            // Styles that we use to hide the labels whilst also keeping it accessible for screen readers
-            const labelStyle = 'overflow:hidden; display:block; width:1px; height:1px';
+            const editorDiv = this.makeEditorDiv();
+            const editorLabel = this.makeLabel('Python Script Area', editorDiv);
+            this.outDiv = this.makeOutDiv();
 
-            // Code editor Label
-            this.editorNode.id = 'code-editor';
-            const editorLabel = document.createElement('label');
-            editorLabel.innerHTML = 'Python Script Area';
-            editorLabel.setAttribute('style', labelStyle);
-            editorLabel.htmlFor = 'code-editor';
+            boxDiv.append(editorLabel);
+            boxDiv.appendChild(editorDiv);
```

Comment:
So, this means we always create the output div, even if we have an explicit `output` attribute?

Reply:
yes, I did it intentionally because it simplify handling the case in which `output=...` points to a non-existent ID:
https://github.com/pyscript/pyscript/blob/02ca66587cebdf7a28fd05cc9e0427da05ce089c/pyscriptjs/src/components/pyrepl.ts#L187-L191

I think that an empty div should not cause problems, does it?

Reply:
I don't think it does and I'm probably too picky here. It just leaves an unused empty tag which I don't think is a big deal.